geth/node/manager.go

```diff
@@ -69,14 +77,108 @@ func (m *NodeManager) startNode(config *params.NodeConfig) (<-chan struct{}, err
 
 	m.initLog(config)
 
-	ethNode, err := MakeNode(config)
+	deliveryManager := new(notifications.DeliveryService)
+
+	ethNode, err := MakeNode(config, deliveryManager)
 	if err != nil {
 		return nil, err
 	}
 
+	m.deliveryService = deliveryManager
 	m.nodeStarted = make(chan struct{}, 1)
 
+	// Subscribe for message delivery status and log out with special key.
+	messageStateLoggingID := deliveryManager.Subscribe(func(state notifications.DeliveryState) {
+		if state.IsP2P {
+			var payload []byte
+			var from, to string
+
+			if state.P2P.Direction == gethmessage.IncomingMessage {
+				if state.P2P.Received != nil {
+					payload = state.P2P.Received.Payload
+
+					if state.P2P.Received.Src != nil {
+						from = gethcommon.ToHex(crypto.FromECDSAPub(state.P2P.Received.Src))
+					}
+
+					if state.P2P.Received.Dst != nil {
+						to = gethcommon.ToHex(crypto.FromECDSAPub(state.P2P.Received.Dst))
+					}
+				}
+			}
+
+			if state.P2P.Direction == gethmessage.OutgoingMessage {
+				from = state.P2P.Source.Sig
+
+				if len(state.P2P.Source.PublicKey) == 0 {
+					to = string(state.P2P.Source.PublicKey)
+				} else {
+					to = state.P2P.Source.TargetPeer
+				}
+			}
+
+			if stat, err := json.Marshal(common.MessageStat{
```

Comment:
It's technologically ok, but prefer single steps for `MessageStat` instance creation, marshalling, and error check. Would reduce complexity for later maintainer.